src/models/popUp.js

```diff
@@ -1,14 +1,150 @@
+                const element = fnode.element;
+                for (const ancestor of ancestors(element)) {
+                    const style = getComputedStyle(ancestor);
+                    if (style.getPropertyValue('visibility') === 'hidden' ||
+                        style.getPropertyValue('display') === 'none') {
+                        return ZEROISH ** coeffVisibleOverlay;
+                    }
+                    // Could add opacity and size checks here, but the
+                    // "nearlyOpaque" and "big" rules already deal with opacity
+                    // and size. If they don't do their jobs, maybe repeat
+                    // their work here (so it gets a different coefficient).
+                }
+                return ONEISH ** coeffVisibleOverlay;
+            }
+            /* Utility procedures */
+
+            /**
+             * Return the extracted [r, g, b, a] values from a string like "rgba(0, 5, 255, 0.8)",
+             * and scale them to 0..1. If no alpha is specified, return undefined for it.
+             */
+            function rgbaFromString(str) {
+                const m = str.match(/^rgba?\s*\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*(?:,\s*(\d+(?:\.\d+)?)\s*)?\)$/i);
+                if (m) {
+                    return [m[1] / 255, m[2] / 255, m[3] / 255, m[4] === undefined ? undefined : parseFloat(m[4])];
+                } else {
+                    throw new Error("Color " + str + " did not match pattern rgb[a](r, g, b[, a]).");
+                }
+            }
+            /**
+             * Return the saturation 0..1 of a color defined by RGB values 0..1.
+             */
+            function saturation(r, g, b) {
+                const cMax = Math.max(r, g, b);
+                const cMin = Math.min(r, g, b);
+                const delta = cMax - cMin;
+                const lightness = (cMax + cMin) / 2;
+                const denom = (1 - (Math.abs(2 * lightness - 1)));
+                // Return 0 if it's black (R, G, and B all 0).
+                return (denom === 0) ? 0 : delta / denom;
+            }
+            function linearScale(number, zeroAt, oneAt) {
+                const isRising = zeroAt < oneAt;
+                if (isRising) {
+                    if (number <= zeroAt) {
+                        return ZEROISH;
+                    }
+                    if (number >= oneAt) {
+                        return ONEISH;
+                    }
+                } else {
+                    if (number >= zeroAt) {
+                        return ZEROISH;
+                    }
+                    if (number <= oneAt) {
+                        return ONEISH;
+                    }
+                }
+                const slope = (ONEISH - ZEROISH) / (oneAt - zeroAt);
+                return slope * (number - zeroAt) + ZEROISH;
+            }
+
+
+            function isNearOverlay(fnode) {
+                const overlayFnode = getHighestScoringOverlay(fnode); 
```

Comment:
On the off chance there are no divs on the page, this will be undefined. You might want to return ZEROISH in this case.